### DISCUSSION TITLE
finetune dpa3

### DISCUSSION THREAD

**Discussion by chtchelkatchev** [deepmodeling/deepmd-kit#4830] [Q&A]
I started using the finetune DPA3, as recommended in https://www.bohrium.com/notebooks/57111746135, for my tasks using the latest deepmd-3.1.0. The fine-tuning appeared to go well externally, and lcurve.out showed high convergence accuracy for all parameters. However, the dp test check revealed that, for example, the energy per atom had an accuracy much worse (by more than an order of magnitude) than what was shown in lcurve.out.

To ensure that this was not my mistake but rather a bug in DeepMD, I tried reproducing the finetune water example suggested in https://www.bohrium.com/notebooks/57111746135. Here, the situation was similar. The lcurve.out matched the one in https://www.bohrium.com/notebooks/57111746135, but dptest showed different results. The dptest values from https://www.bohrium.com/notebooks/57111746135 could not be reproduced:
lcurve.out
# step       rmse_val
    rmse_trn    rmse_e_val  rmse_e_trn    rmse_f_val  rmse_f_trn    rmse_v_val  rmse_v_trn   lr      
# If there is no available reference data, rmse_*_{val,trn} will print nan
      1      2.44e+00    3.09e+00      2.63e-03    6.05e-04      2.44e-01    3.09e-01           nan         nan    1.0e-04
    100      1.42e-01    1.75e-01      1.85e-04    2.25e-03      1.56e-02    1.62e-02           nan         nan    5.5e-05
    200      1.45e-01    1.33e-01      8.88e-04    6.40e-04      1.62e-02    1.52e-02           nan         nan    3.0e-05

my dp test
[2025-07-01 10:22:08,347] DEEPMD INFO    Energy MAE/Natoms  : 1.124378e-03 eV
[2025-07-01 10:22:08,347] DEEPMD INFO    Energy RMSE/Natoms : 1.138879e-03 eV

https://www.bohrium.com/notebooks/57111746135 results
[2025-06-10 15:32:49,768] DEEPMD INFO    Energy MAE/Natoms  : 7.430266e-04 eV
[2025-06-10 15:32:49,768] DEEPMD INFO    Energy RMSE/Natoms : 7.646335e-04 eV

[input_finetune.json](https://github.com/user-attachments/files/20994493/input_finetune.json)
dp --pt train input_finetune.json --skip-neighbor-stat --finetune DPA-3.1-3M.pt --model-branch MP_traj_v024_alldata_mixu   < /dev/null && break

**Comment by iProzd** [comment]
@chtchelkatchev Hi, there might be two primary factors causing the confusion:
1. The values in `lcurve.out` are calculated on-the-fly using randomly sampled individual frames during training, which might not represent the comprehensive accuracy. Please refer to `dp test` which shows the overall accuracy across the entire dataset.
2. The fine-tuning process in the notebook is for quick demonstration purposes with minimal training steps (200), which may bring randomness. Here I reproduced the experiment following the notebook precisely but also obtained different outcomes:

```
dp --pt test -m /deepmd-kit/examples/water/dpa3_finetune/model.ckpt.pt -s /deepmd-kit/examples/water/data
```
First run:
```
[2025-07-03 14:34:45,530] DEEPMD WARNING To get the best performance, it is recommended to adjust the number of threads by setting the environment variables OMP_NUM_THREADS, DP_INTRA_OP_PARALLELISM_THREADS, and DP_INTER_OP_PARALLELISM_THREADS. See https://deepmd.rtfd.io/parallelism/ for more information.
[2025-07-03 14:34:50,402] DEEPMD INFO    # ---------------output of dp test--------------- 
[2025-07-03 14:34:50,403] DEEPMD INFO    # testing system : /deepmd-kit/examples/water/data/data_0
[2025-07-03 14:35:01,861] DEEPMD INFO    Adjust batch size from 1024 to 512
[2025-07-03 14:36:23,435] DEEPMD INFO    # number of test data : 80 
[2025-07-03 14:36:23,436] DEEPMD INFO    Energy MAE         : 3.460127e-02 eV
[2025-07-03 14:36:23,436] DEEPMD INFO    Energy RMSE        : 4.139965e-02 eV
[2025-07-03 14:36:23,436] DEEPMD INFO    Energy MAE/Natoms  : 1.802149e-04 eV
[2025-07-03 14:36:23,436] DEEPMD INFO    Energy RMSE/Natoms : 2.156232e-04 eV
[2025-07-03 14:36:23,436] DEEPMD INFO    Force  MAE         : 1.218762e-02 eV/A
[2025-07-03 14:36:23,436] DEEPMD INFO    Force  RMSE        : 1.588450e-02 eV/A
[2025-07-03 14:36:23,436] DEEPMD INFO    # ----------------------------------------------- 
[2025-07-03 14:36:23,436] DEEPMD INFO    # ---------------output of dp test--------------- 
[2025-07-03 14:36:23,436] DEEPMD INFO    # testing system : /deepmd-kit/examples/water/data/data_1
[2025-07-03 14:36:43,719] DEEPMD INFO    # number of test data : 160 
[2025-07-03 14:36:43,719] DEEPMD INFO    Energy MAE         : 3.727699e-02 eV
[2025-07-03 14:36:43,719] DEEPMD INFO    Energy RMSE        : 4.406429e-02 eV
[2025-07-03 14:36:43,719] DEEPMD INFO    Energy MAE/Natoms  : 1.941510e-04 eV
[2025-07-03 14:36:43,719] DEEPMD INFO    Energy RMSE/Natoms : 2.295015e-04 eV
[2025-07-03 14:36:43,719] DEEPMD INFO    Force  MAE         : 1.220880e-02 eV/A
[2025-07-03 14:36:43,719] DEEPMD INFO    Force  RMSE        : 1.585310e-02 eV/A
[2025-07-03 14:36:43,727] DEEPMD INFO    # ----------------------------------------------- 
[2025-07-03 14:36:43,727] DEEPMD INFO    # ---------------output of dp test--------------- 
[2025-07-03 14:36:43,727] DEEPMD INFO    # testing system : /deepmd-kit/examples/water/data/data_3
[2025-07-03 14:36:53,874] DEEPMD INFO    # number of test data : 80 
[2025-07-03 14:36:53,874] DEEPMD INFO    Energy MAE         : 3.108779e-02 eV
[2025-07-03 14:36:53,874] DEEPMD INFO    Energy RMSE        : 3.879563e-02 eV
[2025-07-03 14:36:53,874] DEEPMD INFO    Energy MAE/Natoms  : 1.619156e-04 eV
[2025-07-03 14:36:53,874] DEEPMD INFO    Energy RMSE/Natoms : 2.020606e-04 eV
[2025-07-03 14:36:53,874] DEEPMD INFO    Force  MAE         : 1.239506e-02 eV/A
[2025-07-03 14:36:53,874] DEEPMD INFO    Force  RMSE        : 1.605893e-02 eV/A
[2025-07-03 14:36:53,875] DEEPMD INFO    # ----------------------------------------------- 
[2025-07-03 14:36:53,875] DEEPMD INFO    # ---------------output of dp test--------------- 
[2025-07-03 14:36:53,875] DEEPMD INFO    # testing system : /deepmd-kit/examples/water/data/data_2
[2025-07-03 14:37:04,015] DEEPMD INFO    # number of test data : 80 
[2025-07-03 14:37:04,015] DEEPMD INFO    Energy MAE         : 3.413328e-02 eV
[2025-07-03 14:37:04,015] DEEPMD INFO    Energy RMSE        : 4.206205e-02 eV
[2025-07-03 14:37:04,015] DEEPMD INFO    Energy MAE/Natoms  : 1.777775e-04 eV
[2025-07-03 14:37:04,015] DEEPMD INFO    Energy RMSE/Natoms : 2.190732e-04 eV
[2025-07-03 14:37:04,015] DEEPMD INFO    Force  MAE         : 1.215139e-02 eV/A
[2025-07-03 14:37:04,015] DEEPMD INFO    Force  RMSE        : 1.578798e-02 eV/A
[2025-07-03 14:37:04,015] DEEPMD INFO    # ----------------------------------------------- 
[2025-07-03 14:37:04,015] DEEPMD INFO    # ----------weighted average of errors----------- 
[2025-07-03 14:37:04,015] DEEPMD INFO    # number of systems : 4
[2025-07-03 14:37:04,015] DEEPMD INFO    Energy MAE         : 3.487526e-02 eV
[2025-07-03 14:37:04,015] DEEPMD INFO    Energy RMSE        : 4.212261e-02 eV
[2025-07-03 14:37:04,015] DEEPMD INFO    Energy MAE/Natoms  : 1.816420e-04 eV
[2025-07-03 14:37:04,015] DEEPMD INFO    Energy RMSE/Natoms : 2.193886e-04 eV
[2025-07-03 14:37:04,015] DEEPMD INFO    Force  MAE         : 1.223033e-02 eV/A
[2025-07-03 14:37:04,015] DEEPMD INFO    Force  RMSE        : 1.588778e-02 eV/A
[2025-07-03 14:37:04,015] DEEPMD INFO    Virial MAE         : 0.000000e+00 eV
[2025-07-03 14:37:04,016] DEEPMD INFO    Virial RMSE        : 0.000000e+00 eV
[2025-07-03 14:37:04,016] DEEPMD INFO    Virial MAE/Natoms  : 0.000000e+00 eV
[2025-07-03 14:37:04,016] DEEPMD INFO    Virial RMSE/Natoms : 0.000000e+00 eV
[2025-07-03 14:37:04,016] DEEPMD INFO    # ----------------------------------------------- 
```
 Second run following exactly the same process:

```
[2025-07-03 14:39:00,506] DEEPMD WARNING To get the best performance, it is recommended to adjust the number of threads by setting the environment variables OMP_NUM_THREADS, DP_INTRA_OP_PARALLELISM_THREADS, and DP_INTER_OP_PARALLELISM_THREADS. See https://deepmd.rtfd.io/parallelism/ for more information.
[2025-07-03 14:39:05,371] DEEPMD INFO    # ---------------output of dp test--------------- 
[2025-07-03 14:39:05,371] DEEPMD INFO    # testing system : /deepmd-kit/examples/water/data/data_0
[2025-07-03 14:39:16,667] DEEPMD INFO    Adjust batch size from 1024 to 512
[2025-07-03 14:40:36,642] DEEPMD INFO    # number of test data : 80 
[2025-07-03 14:40:36,643] DEEPMD INFO    Energy MAE         : 1.555517e-01 eV
[2025-07-03 14:40:36,643] DEEPMD INFO    Energy RMSE        : 1.595196e-01 eV
[2025-07-03 14:40:36,643] DEEPMD INFO    Energy MAE/Natoms  : 8.101650e-04 eV
[2025-07-03 14:40:36,643] DEEPMD INFO    Energy RMSE/Natoms : 8.308314e-04 eV
[2025-07-03 14:40:36,643] DEEPMD INFO    Force  MAE         : 1.205806e-02 eV/A
[2025-07-03 14:40:36,643] DEEPMD INFO    Force  RMSE        : 1.570110e-02 eV/A
[2025-07-03 14:40:36,643] DEEPMD INFO    # ----------------------------------------------- 
[2025-07-03 14:40:36,643] DEEPMD INFO    # ---------------output of dp test--------------- 
[2025-07-03 14:40:36,643] DEEPMD INFO    # testing system : /deepmd-kit/examples/water/data/data_1
[2025-07-03 14:40:56,943] DEEPMD INFO    # number of test data : 160 
[2025-07-03 14:40:56,943] DEEPMD INFO    Energy MAE         : 1.554817e-01 eV
[2025-07-03 14:40:56,943] DEEPMD INFO    Energy RMSE        : 1.602297e-01 eV
[2025-07-03 14:40:56,943] DEEPMD INFO    Energy MAE/Natoms  : 8.098006e-04 eV
[2025-07-03 14:40:56,943] DEEPMD INFO    Energy RMSE/Natoms : 8.345298e-04 eV
[2025-07-03 14:40:56,943] DEEPMD INFO    Force  MAE         : 1.204150e-02 eV/A
[2025-07-03 14:40:56,943] DEEPMD INFO    Force  RMSE        : 1.563364e-02 eV/A
[2025-07-03 14:40:56,951] DEEPMD INFO    # ----------------------------------------------- 
[2025-07-03 14:40:56,951] DEEPMD INFO    # ---------------output of dp test--------------- 
[2025-07-03 14:40:56,951] DEEPMD INFO    # testing system : /deepmd-kit/examples/water/data/data_3
[2025-07-03 14:41:07,108] DEEPMD INFO    # number of test data : 80 
[2025-07-03 14:41:07,108] DEEPMD INFO    Energy MAE         : 1.507844e-01 eV
[2025-07-03 14:41:07,108] DEEPMD INFO    Energy RMSE        : 1.549291e-01 eV
[2025-07-03 14:41:07,108] DEEPMD INFO    Energy MAE/Natoms  : 7.853354e-04 eV
[2025-07-03 14:41:07,108] DEEPMD INFO    Energy RMSE/Natoms : 8.069227e-04 eV
[2025-07-03 14:41:07,108] DEEPMD INFO    Force  MAE         : 1.223306e-02 eV/A
[2025-07-03 14:41:07,108] DEEPMD INFO    Force  RMSE        : 1.583778e-02 eV/A
[2025-07-03 14:41:07,108] DEEPMD INFO    # ----------------------------------------------- 
[2025-07-03 14:41:07,108] DEEPMD INFO    # ---------------output of dp test--------------- 
[2025-07-03 14:41:07,109] DEEPMD INFO    # testing system : /deepmd-kit/examples/water/data/data_2
[2025-07-03 14:41:17,246] DEEPMD INFO    # number of test data : 80 
[2025-07-03 14:41:17,246] DEEPMD INFO    Energy MAE         : 1.613958e-01 eV
[2025-07-03 14:41:17,246] DEEPMD INFO    Energy RMSE        : 1.646936e-01 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Energy MAE/Natoms  : 8.406031e-04 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Energy RMSE/Natoms : 8.577794e-04 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Force  MAE         : 1.195881e-02 eV/A
[2025-07-03 14:41:17,247] DEEPMD INFO    Force  RMSE        : 1.554244e-02 eV/A
[2025-07-03 14:41:17,247] DEEPMD INFO    # ----------------------------------------------- 
[2025-07-03 14:41:17,247] DEEPMD INFO    # ----------weighted average of errors----------- 
[2025-07-03 14:41:17,247] DEEPMD INFO    # number of systems : 4
[2025-07-03 14:41:17,247] DEEPMD INFO    Energy MAE         : 1.557391e-01 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Energy RMSE        : 1.599504e-01 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Energy MAE/Natoms  : 8.111410e-04 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Energy RMSE/Natoms : 8.330751e-04 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Force  MAE         : 1.206659e-02 eV/A
[2025-07-03 14:41:17,247] DEEPMD INFO    Force  RMSE        : 1.567003e-02 eV/A
[2025-07-03 14:41:17,247] DEEPMD INFO    Virial MAE         : 0.000000e+00 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Virial RMSE        : 0.000000e+00 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Virial MAE/Natoms  : 0.000000e+00 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    Virial RMSE/Natoms : 0.000000e+00 eV
[2025-07-03 14:41:17,247] DEEPMD INFO    # ----------------------------------------------- 
```

**Reply by chtchelkatchev** [reply]
Thank you for the quick response. Given the observed differences when running finetuning on a small database, I suspect that finetuning affects not only the outer layers of the fitting net but also the graph descriptor. This is supported by the out.json file, where the descriptor is marked as "trainable": true. Could you briefly clarify which parts of the DPA-3 potential are actually trained during finetuning and which remain frozen? Is there any way to manually control what gets frozen?

On a related note, the developers of NEQUIP claim that even 100 configurations can sometimes be sufficient to train NEQUIP from scratch, as demonstrated in their examples. However, it appears that with the same number of configurations, deepmd-dpa3 cannot even be finetuned effectively.

**Reply by anyangml** [reply]
By default, it's a full finetuning. However, you can freeze the descriptor if you want, see [doc](https://docs.deepmodeling.com/projects/deepmd/en/r2/train/finetuning.html).
